### DISCUSSION TITLE
How to optimistically update an infinite or paginated query?

### DISCUSSION THREAD

**Discussion by longprao** [TanStack/query#848] [General]
I have read the docs and the repo issues several times but still can't figure out how to use `queryCache.setQueryData()` to add/update/remove items from an `useInfiniteQuery()`? Not sure what to pass to the first params of `setQueryData`?

It seems for now on mutation success I always have to call `queryCache.invalidateQueries()` which is very wasteful.

Some code example would be great. Thanks!

**Comment by criticalx7** [comment]
For now what I did was passing a query group index to a mutation. Now i know which query group of an item that need to be updated belongs. Next, I looped through that query group to find that will-be-updated item (Immerjs helps a lot). The variables is available as argument on all on-prefix callback. Hope it helps you. 

I kind of wish the the dev guide us in someway on this topic.  My way is kind of usable and demo-able but it feels not scalable somehow.

@tannerlinsley  Can you help guide us in proper way for this problem ?

**Reply by EQuimper** [reply]
Did the same. Immer helps a lot. I just hope the user dont gonna scroll 100 pages haha.

**Comment by ranyefet** [comment]
I'm also having the same issue.

I use infinite list with potential lots of items and allow to select a few and bulk update them.
I need a way to update them optimistically without re-fetching all the pages again.

What is the best way to achieve this?

**Comment by youbek** [comment]
Have you found a solution, guys?

**Comment by TkDodo** [comment]
since v3, useInfiniteQuery consists of an object with `pages` and `pageParams`:

```
export interface InfiniteData<TData> {
  pages: TData[]
  pageParams: unknown[]
}
```

given that I have an array of todos on each page, I would use `setQueryData` to update the name of a todo like this:

```
setQueryData(data => ({
    ...data,
    pages: data.pages.map(page => page.map(todo => todo.id === id ? { ...todo, name: 'new name' } : todo ))
})
```

**Reply by kelsman** [reply]
This didn't work for me

**Reply by TkDodo** [reply]
Then please provide a codesandbox example that shows that it’s not working.

**Reply by kelsman** [reply]
i wrote this on an onSuccess call back
   queryClient.setQueryData('feedsData', prev => {
        return {
          ...prev,
          pages: prev.pages.map(page => page.posts.map(post => post._id === feed._id ? { ...post, comments: [...post.comments, data.newComment] } : post))
        }
      })

**Reply by kelsman** [reply]
i get a  "cannot read property map of undefined" the undefined in question here is the data gotten from feedsData useInfinitequery

**Reply by TkDodo** [reply]
well that just means that `prev` is undefined, which happens when there is no data in the cache for the key `'feedsData'` ...

**Reply by kelsman** [reply]
Console logging "prev" returns the initially fetched "feeds data"

**Reply by TkDodo** [reply]
sorry, I'm not gonna keep playing a guessing game here :) As I said before:

>  please provide a codesandbox example that shows that it’s not working.

**Reply by kelsman** [reply]
sorry about that, here you go.
https://codesandbox.io/s/hardcore-cohen-cdons?file=/src/card.js

**Reply by kelsman** [reply]
try commenting on a picture

**Reply by TkDodo** [reply]
a simple logging shows that you are changing the structure of the data in the cache.

initially, every `page` has a `posts` array and a `next` number. After the update, each entry in the `pages` array contains only `_id`, `author` etc ...

in your update in `onSuccess`, you are not spreading `...page`. here is a fixed version:

https://codesandbox.io/s/bold-river-0e6qb

- `TypeScript` helps tremendously in spotting such things
- `immer` is very good for updating deeply nested state in an immutable way, without needing to spread everything manually.

**Reply by kelsman** [reply]
thank you so much for the assistance.

**Comment by apenab** [comment]
With DRF I have this solution:

```js
export function useInfiniteQuery<
  TQueryFnData = unknown,
  TError = unknown,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = QueryKey
>(
  queryKey: TQueryKey,
  baseUrl: string,
  params?: Record<string, string>,
  options?: Omit<UseInfiniteQueryOptions<TQueryFnData, TError, TData, TQueryFnData, TQueryKey>, "queryKey" | "queryFn">
) {
  const infiniteQueryObject = _useInfiniteQuery<TQueryFnData, TError, TData, TQueryKey>(
    queryKey,
    async ({pageParam}) => {
      const response = await axios.get(pageParam || baseUrl, !pageParam ? {params} : {});
      return response.data;
    },
    {
      // TODO: ppr-refactor
      getNextPageParam: (lastPage: any) => lastPage?.next,
      ...options,
    }
  );

  const queryClient = useQueryClient();

  function updateQueryData(newData: TData) {
    queryClient.setQueryData<InfiniteData<PageData<TData>> | undefined>(queryKey, (data) => {
      if (data) {
        data.pages = data.pages.map((page) => {
          page.results = page.results.map((result) => {
            // @ts-ignore
            if (result.id === newData.id) {
              result = newData;
            }
            return result;
          });
          return page;
        });
      }
      return data;
    });
  }

  return {items: transformQueryData<TData>(infiniteQueryObject.data), updateQueryData, ...infiniteQueryObject};
}
```

**Reply by afraser** [reply]
Nailed it. Although for me (using version 5.59.15) I just need `<InfiniteData<TData>> | undefined` for the typing.

Future travelers who do this sort of thing a lot might find they'd benefit from something like [immer](https://immerjs.github.io/immer/) to save them from all the maps and spreads.

**Comment by tkmin** [comment]
is there other better solution than using map func in map func?

**Reply by apenab** [reply]
you can use the fp lodash utilities

**Reply by afraser** [reply]
also [immer](https://immerjs.github.io/immer/)

**Comment by lemuelflores** [comment]
I happen to share my solution to this on another discussion with the same subject (#3360).

This and the said discussion are the top 2 results on Google so I would like share [this](https://github.com/TanStack/query/discussions/3360#discussioncomment-5921065) here as well.
